### DISCUSSION TITLE
How Thanos Querier works?

### DISCUSSION THREAD

**Discussion by yliu138repo** [thanos-io/thanos#7049] [Questions & Answers]
Hi Team,

I am new to Thanos and Prometheus. Thanos is a great project, and helps to build robust monitoring systems. So we are keen on gaining more insights into this product.

Our current prometehus instance is deployed into on-prem infrastructure, and after integrating with Thanos, we might want to upload the TSDB blocks to GCS. Since connecting to public cloud will experience network latency, we'd like to know what impact will this be than directly query via prometheus locally for historical data, and what is suggested configuration for query or any other components to avoid obvious latency?

Another question is that how does Thanos query handles the blocks that has been uploaded to the remote storage, which are still available in the Promtheus locally (still within the retentions period configured)? does the querier read from the sidecar still  to reduce latency. Also how about the case when query happens and one blocks is still been transferred? 

 If the code snippet could be provided, that would be greatly aprpeciated.

Thanks in advance.

**Comment by MichaHoffmann** [comment]
Generally querier will fan-out storage requests to its configured "endpoints" (after pruning for only those that could return results based on range of the query, and external labels and such) and then deduplicate and merge the results and execute the PromQL query on this set of merged results.
The component that queries data from GCS is the thanos `store`; it heavily caches postings ~ for us latency has not been much of an issue so far.
